collector/test/ConnTrackerTest.cpp

```diff
@@ -1246,91 +1246,149 @@ TEST(ConnTrackerTest, TestComputeDeltaAfterglowBenchmark2) {
+
+ private:
+  uint64_t pid_;
+  std::string container_id_;
+  std::string comm_;  // part of the process-unique-key
+  std::string exe_;
+  std::string exe_path_;  // part of the process-unique-key
+  std::string args_;      // part of the process-unique-key
+};
+
+/* Same endpoint, but opened by two processes having different process-unique-key.
+   We expect that the endpoint is reported once for each of the processes. */
+TEST(ConnTrackerTest, TestSameEndpointDifferentProcess) {
+  int64_t activity_time1 = 1000;
+  int64_t activity_time2 = 2000;
   Endpoint a(Address(192, 168, 0, 1), 80);
   ConnectionTracker connectionTracker;
-  std::shared_ptr<Process> process1 = std::make_shared<Process>(2);
-  std::shared_ptr<Process> process2 = std::make_shared<Process>(3);
+  std::shared_ptr<IProcess> process1 = std::make_shared<FakeProcess>(2, "container", "specific_comm1", "exe", "exe_path", "args");
+  std::shared_ptr<IProcess> process2 = std::make_shared<FakeProcess>(3, "container", "specific_comm2", "exe", "exe_path", "args");
 
-  ContainerEndpoint ce1(container, a, L4Proto::TCP, process1);
-  ContainerEndpoint ce2(container, a, L4Proto::TCP, process2);
-
-  ConnStatus oldConnStatus(connection_time1, true);
-  ConnStatus newConnStatus(connection_time2, true);
+  ContainerEndpoint ce1("container", a, L4Proto::TCP, process1);
+  ContainerEndpoint ce2("container", a, L4Proto::TCP, process2);
 
-  connectionTracker.EmplaceOrUpdateNoLock(ce1, oldConnStatus);
-  connectionTracker.EmplaceOrUpdateNoLock(ce2, newConnStatus);
+  ConnStatus connStatus1(activity_time1, true);
+  ConnStatus connStatus2(activity_time2, true);
 
-  ContainerEndpointMap expected_state = {{ce1, oldConnStatus}, {ce2, newConnStatus}};
-  ContainerEndpointMap observed_state = connectionTracker.FetchEndpointState(false, false);
-  EXPECT_THAT(observed_state, expected_state);
+  connectionTracker.EmplaceOrUpdateNoLock(ce1, connStatus1);
+  connectionTracker.EmplaceOrUpdateNoLock(ce2, connStatus2);
 
-  ContainerEndpoint observed_endpoint = observed_state.find(ce2)->first;
-  int observedPid = observed_endpoint.originator()->pid();
+  AdvertisedEndpointMap observed_state = connectionTracker.FetchEndpointState(false, false);
 
-  EXPECT_THAT(observedPid, 3);
+  EXPECT_THAT(observed_state, UnorderedElementsAre(std::make_pair(ce1, connStatus1), std::make_pair(ce2, connStatus2)));
 }
+/* Same endpoint, but opened by two processes having the same process-unique-key.
+   We expect that the endpoint is reported only once */
+TEST(ConnTrackerTest, TestSameEndpointSimilarProcess) {
+  int64_t activity_time1 = 1000;
+  int64_t activity_time2 = 2000;
+  Endpoint a(Address(192, 168, 0, 1), 80);
+  ConnectionTracker connectionTracker;
+  std::shared_ptr<IProcess> process1 = std::make_shared<FakeProcess>(2, "container", "comm", "exe1", "exe_path", "args");
+  std::shared_ptr<IProcess> process2 = std::make_shared<FakeProcess>(3, "container", "comm", "exe2", "exe_path", "args");
```

Comment:
Write a unit test where there are two processes with the same pid. This would help to show that what is being compared is the addresses of the shared pointers rather than their contents.

Reply:
Unit tests are useful for their expression of functional expectations; I don't think that they should fail when some internal implementation details are changed. Although, we might want to make sure that the ProcessStore always returns the same pointer for a given Process object.

Reply:
Okay